akka-actor/src/main/scala/akka/actor/ActorRef.scala

```diff
@@ -732,83 +732,106 @@ private[akka] final class FunctionRef(
     }
   }
 
-  // requires sychronized access because AddressTerminatedTopic must be updated together with this
+  // watching, _watchedBy and maintainAddressTerminatedSubscription requires sychronized access because
+  // AddressTerminatedTopic must be updated together with the variables here.
+  // Important: don't include calls to sendSystemMessage inside the synchronized since that can
+  // result in deadlock, see issue #26326
   private[this] var watching = ActorCell.emptyActorRefSet
-  // requires sychronized access because AddressTerminatedTopic must be updated together with this
   private[this] var _watchedBy: OptionVal[Set[ActorRef]] = OptionVal.Some(ActorCell.emptyActorRefSet)
 
   override def isTerminated: Boolean = _watchedBy.isEmpty
 
   //noinspection EmptyCheck
-  protected def sendTerminated(): Unit = synchronized {
+  protected def sendTerminated(): Unit = {
     def unwatchWatched(watched: ActorRef): Unit =
       watched.asInstanceOf[InternalActorRef].sendSystemMessage(Unwatch(watched, this))
 
-    _watchedBy match {
-      case OptionVal.Some(watchedBy) ⇒
-        if (watchedBy.nonEmpty) {
-          watchedBy foreach sendTerminated(ifLocal = false)
-          watchedBy foreach sendTerminated(ifLocal = true)
-        }
+    val toUnwatch = this.synchronized {
+      _watchedBy match {
+        case OptionVal.Some(watchedBy) ⇒
+          if (watchedBy.nonEmpty) {
+            watchedBy foreach sendTerminated(ifLocal = false)
```

Comment:
Are these two still ok to be inside the synchronized block?

Reply:
good catch, my bad

Reply:
fixed in 4310184